File: src/learn/tree/PathSum_II.java
package learn.tree;

import java.util.ArrayList;
import java.util.List;

import helper.Helper;
import models.TreeNodeJava;

// * Time: O(n logn)
//   ? Length of slate for one leaf node could be (log n) in the worst case, when all paths to leaf == target
// * Space: O(n logn)
public class PathSum_II {
    public static void main(String[] args) {
        System.out.println(pathSum_II(Helper.createBinaryTree(), 15));
    }

    private static List<List<Integer>> pathSum_II(TreeNodeJava root, Integer target) {
        if (root == null)
            return null;

        List<List<Integer>> result = new ArrayList<>();
        dfsHelper(root, target, new ArrayList<Integer>(), result);
        return result;
    }

    private static void dfsHelper(TreeNodeJava node, Integer target, List<Integer> slate, List<List<Integer>> result) {
        // Base case
        if (node.left == null && node.right == null) {
            if (target == node.val) {
                slate.add(node.val);
                result.add(new ArrayList<>(slate));
                slate.remove(slate.size() - 1);
            }
            return;
        }
        // Recursive case
        slate.add(node.val);
        if (node.left != null) {
            dfsHelper(node.left, target - node.val, slate, result);
        }
        if (node.right != null) {
            dfsHelper(node.right, target - node.val, slate, result);
        }
        slate.remove(slate.size() - 1);
    }
}

File: src/learn/tree/PathSumII.kt
package learn.tree

import models.TreeNode
import java.util.ArrayList

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */
internal class PathSumII {
    private fun recurseTree(
        node: TreeNode?,
        remainingSum: Int,
        pathNodes: MutableList<Int>,
        pathsList: MutableList<List<Int>>
    ) {
        if (node == null) {
            return
        }

        // Add the current node to the path's list
        pathNodes.add(node.`val`)

        // Check if the current node is a leaf and also, if it
        // equals our remaining sum. If it does, we add the path to
        // our list of paths
        if (remainingSum == node.`val` && node.left == null && node.right == null) {
            pathsList.add(ArrayList(pathNodes))
        } else {

            // Else, we will recurse on the left and the right children
            recurseTree(node.left, remainingSum - node.`val`, pathNodes, pathsList)
            recurseTree(node.right, remainingSum - node.`val`, pathNodes, pathsList)
        }

        // We need to pop the node once we are done processing ALL of it's
        // subtrees.
        pathNodes.removeAt(pathNodes.size - 1)
    }

    fun pathSum(root: TreeNode?, sum: Int): List<List<Int>> {
        val pathsList: MutableList<List<Int>> = ArrayList()
        val pathNodes: MutableList<Int> = ArrayList()
        recurseTree(root, sum, pathNodes, pathsList)
        return pathsList
    }
}